File: 999.available-captures-for-rook.cpp
/*
 * @lc app=leetcode id=999 lang=cpp
 *
 * [999] Available Captures for Rook
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  int numRookCaptures(vector<vector<char>>& board) {
    int rook_x = 0;
    int rook_y = 0;
    for (int y = 0; y < 8; y++) {
      for (int x = 0; x < 8; x++) {
        if (board[y][x] == 'R') {
          rook_x = x;
          rook_y = y;
          break;
        }
      }
    }
    int ret = 0;
    int directions[][2] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    for (const auto& dir : directions) {
      int dx = dir[0];
      int dy = dir[1];
      int x = rook_x + dx;
      int y = rook_y + dy;
      while (y >= 0 && y < 8 && x >= 0 && x < 8) {
        char val = board[y][x];
        if (val == 'B') {
          break;
        }
        if (val == 'p') {
          ret++;
          break;
        }
        x += dx;
        y += dy;
      }
    }
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<vector<char>> board;

  board = {{'.', '.', '.', '.', '.', '.', '.', '.'},
           {'.', '.', '.', 'p', '.', '.', '.', '.'},
           {'.', '.', '.', 'R', '.', '.', '.', 'p'},
           {'.', '.', '.', '.', '.', '.', '.', '.'},
           {'.', '.', '.', '.', '.', '.', '.', '.'},
           {'.', '.', '.', 'p', '.', '.', '.', '.'},
           {'.', '.', '.', '.', '.', '.', '.', '.'},
           {'.', '.', '.', '.', '.', '.', '.', '.'}};
  ASSERT_EQ(s.numRookCaptures(board), 3);

  board = {{'.', '.', '.', '.', '.', '.', '.', '.'},
           {'.', 'p', 'p', 'p', 'p', 'p', '.', '.'},
           {'.', 'p', 'p', 'B', 'p', 'p', '.', '.'},
           {'.', 'p', 'B', 'R', 'B', 'p', '.', '.'},
           {'.', 'p', 'p', 'B', 'p', 'p', '.', '.'},
           {'.', 'p', 'p', 'p', 'p', 'p', '.', '.'},
           {'.', '.', '.', '.', '.', '.', '.', '.'},
           {'.', '.', '.', '.', '.', '.', '.', '.'}};
  ASSERT_EQ(s.numRookCaptures(board), 0);

  board = {{'.', '.', '.', '.', '.', '.', '.', '.'},
           {'.', '.', '.', 'p', '.', '.', '.', '.'},
           {'.', '.', '.', 'p', '.', '.', '.', '.'},
           {'p', 'p', '.', 'R', '.', 'p', 'B', '.'},
           {'.', '.', '.', '.', '.', '.', '.', '.'},
           {'.', '.', '.', 'B', '.', '.', '.', '.'},
           {'.', '.', '.', 'p', '.', '.', '.', '.'},
           {'.', '.', '.', '.', '.', '.', '.', '.'}};
  ASSERT_EQ(s.numRookCaptures(board), 3);

  return 0;
}

File: 999.available-captures-for-rook.rs
/*
 * @lc app=leetcode id=999 lang=rust
 *
 * [999] Available Captures for Rook
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn num_rook_captures(board: Vec<Vec<char>>) -> i32 {
        for i in 0..board.len() {
            for j in 0..board[i].len() {
                if board[i][j] == 'R' {
                    return Solution::cap(&board, i as i32, j as i32, 0, 1)
                        + Solution::cap(&board, i as i32, j as i32, 0, -1)
                        + Solution::cap(&board, i as i32, j as i32, 1, 0)
                        + Solution::cap(&board, i as i32, j as i32, -1, 0);
                }
            }
        }
        0
    }

    fn cap(b: &Vec<Vec<char>>, mut x: i32, mut y: i32, dx: i32, dy: i32) -> i32 {
        while x >= 0
            && x < b.len() as i32
            && y >= 0
            && y < b[x as usize].len() as i32
            && b[x as usize][y as usize] != 'B'
        {
            if b[x as usize][y as usize] == 'p' {
                return 1;
            }
            x += dx;
            y += dy;
        }
        0
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::num_rook_captures(vec![
            vec!['.', '.', '.', '.', '.', '.', '.', '.'],
            vec!['.', '.', '.', 'p', '.', '.', '.', '.'],
            vec!['.', '.', '.', 'R', '.', '.', '.', 'p'],
            vec!['.', '.', '.', '.', '.', '.', '.', '.'],
            vec!['.', '.', '.', '.', '.', '.', '.', '.'],
            vec!['.', '.', '.', 'p', '.', '.', '.', '.'],
            vec!['.', '.', '.', '.', '.', '.', '.', '.'],
            vec!['.', '.', '.', '.', '.', '.', '.', '.']
        ]),
        3
    );
}

#[test]
fn test2() {
    assert_eq!(
        Solution::num_rook_captures(vec![
            vec!['.', '.', '.', '.', '.', '.', '.', '.'],
            vec!['.', 'p', 'p', 'p', 'p', 'p', '.', '.'],
            vec!['.', 'p', 'p', 'B', 'p', 'p', '.', '.'],
            vec!['.', 'p', 'B', 'R', 'B', 'p', '.', '.'],
            vec!['.', 'p', 'p', 'B', 'p', 'p', '.', '.'],
            vec!['.', 'p', 'p', 'p', 'p', 'p', '.', '.'],
            vec!['.', '.', '.', '.', '.', '.', '.', '.'],
            vec!['.', '.', '.', '.', '.', '.', '.', '.']
        ]),
        0
    );
}

#[test]
fn test3() {
    assert_eq!(
        Solution::num_rook_captures(vec![
            vec!['.', '.', '.', '.', '.', '.', '.', '.'],
            vec!['.', '.', '.', 'p', '.', '.', '.', '.'],
            vec!['.', '.', '.', 'p', '.', '.', '.', '.'],
            vec!['p', 'p', '.', 'R', '.', 'p', 'B', '.'],
            vec!['.', '.', '.', '.', '.', '.', '.', '.'],
            vec!['.', '.', '.', 'B', '.', '.', '.', '.'],
            vec!['.', '.', '.', 'p', '.', '.', '.', '.'],
            vec!['.', '.', '.', '.', '.', '.', '.', '.']
        ]),
        3
    );
}
